src/runtime/alloc.cpp

```diff
@@ -448,6 +458,17 @@ void * alloc(size_t sz) {
     return lean_alloc_small(sz, slot_idx);
 }
 
+LEAN_NOINLINE
+static void dealloc_small_core_cold(void * o) {
+    set_next_obj(o, g_heap->m_to_export_list);
+    g_heap->m_to_export_list = o;
+    g_heap->m_to_export_list_size++;
+    if (g_heap->m_to_export_list_size > LEAN_MAX_TO_EXPORT_OBJS) {
+        LEAN_RUNTIME_STAT_CODE(g_num_exports++);
+        g_heap->export_objs();
+    }
+}
+
 static inline void dealloc_small_core(void * o) {
     LEAN_RUNTIME_STAT_CODE(g_num_small_dealloc++);
     if (LEAN_UNLIKELY(g_heap == nullptr)) {
```

Comment:
@leodemoura Does this dynamic check exist for using Lean libraries from non-Lean code? Since for every program/thread started from Lean we eagerly initialize the heap I believe.
Even if the branch is perfectly predictable, we might want to check whether it still incurs some overhead, in which case we could make heap initialization the external user's burden, or possibly that of the Lean module initialization functions.

Reply:
If I remember correctly the test above had two motivations. The one you pointed out, and C++ static initialization code. Both scenarios are not gone. The former may reappear in the future, but we can always instruct users to explicitly call an initialization function.